### ISSUE TITLE
log4js@2.3.5 typings break the build

### ISSUE THREAD

**Issue by mitchhentges** [monounity/karma-typescript#197]
When I clone locally or build on Travis (See [here](https://travis-ci.org/monounity/karma-typescript/jobs/287112740)), my build fails due to `log4js` types:

```
src/api/transforms.ts(2,25): error TS7016: Could not find a declaration file for module 'log4js'. '/home/mitch/dev/karma-typescript/node_modules/log4js/lib/log4js.js' implicitly has an 'any' type.
  Try `npm install @types/log4js` if it exists or add a new declaration (.d.ts) file containing `declare module 'log4js';`
src/bundler/bundler.ts(8,24): error TS7016: Could not find a declaration file for module 'log4js'. '/home/mitch/dev/karma-typescript/node_modules/log4js/lib/log4js.js' implicitly has an 'any' type.
  Try `npm install @types/log4js` if it exists or add a new declaration (.d.ts) file containing `declare module 'log4js';`
src/bundler/dependency-walker.ts(10,24): error TS7016: Could not find a declaration file for module 'log4js'. '/home/mitch/dev/karma-typescript/node_modules/log4js/lib/log4js.js' implicitly has an 'any' type.
  Try `npm install @types/log4js` if it exists or add a new declaration (.d.ts) file containing `declare module 'log4js';`
src/bundler/resolve/resolver.ts(7,24): error TS7016: Could not find a declaration file for module 'log4js'. '/home/mitch/dev/karma-typescript/node_modules/log4js/lib/log4js.js' implicitly has an 'any' type.
  Try `npm install @types/log4js` if it exists or add a new declaration (.d.ts) file containing `declare module 'log4js';`
src/bundler/resolve/source-reader.spec.ts(1,25): error TS7016: Could not find a declaration file for module 'log4js'. '/home/mitch/dev/karma-typescript/node_modules/log4js/lib/log4js.js' implicitly has an 'any' type.
  Try `npm install @types/log4js` if it exists or add a new declaration (.d.ts) file containing `declare module 'log4js';`
src/bundler/resolve/source-reader.ts(6,24): error TS7016: Could not find a declaration file for module 'log4js'. '/home/mitch/dev/karma-typescript/node_modules/log4js/lib/log4js.js' implicitly has an 'any' type.
  Try `npm install @types/log4js` if it exists or add a new declaration (.d.ts) file containing `declare module 'log4js';`
src/bundler/source-map.ts(6,24): error TS7016: Could not find a declaration file for module 'log4js'. '/home/mitch/dev/karma-typescript/node_modules/log4js/lib/log4js.js' implicitly has an 'any' type.
  Try `npm install @types/log4js` if it exists or add a new declaration (.d.ts) file containing `declare module 'log4js';`
src/compiler/compiler.ts(4,24): error TS7016: Could not find a declaration file for module 'log4js'. '/home/mitch/dev/karma-typescript/node_modules/log4js/lib/log4js.js' implicitly has an 'any' type.
  Try `npm install @types/log4js` if it exists or add a new declaration (.d.ts) file containing `declare module 'log4js';`
src/index.ts(1,25): error TS7016: Could not find a declaration file for module 'log4js'. '/home/mitch/dev/karma-typescript/node_modules/log4js/lib/log4js.js' implicitly has an 'any' type.
  Try `npm install @types/log4js` if it exists or add a new declaration (.d.ts) file containing `declare module 'log4js';`
src/istanbul/coverage.ts(1,24): error TS7016: Could not find a declaration file for module 'log4js'. '/home/mitch/dev/karma-typescript/node_modules/log4js/lib/log4js.js' implicitly has an 'any' type.
  Try `npm install @types/log4js` if it exists or add a new declaration (.d.ts) file containing `declare module 'log4js';`
src/istanbul/threshold.ts(6,24): error TS7016: Could not find a declaration file for module 'log4js'. '/home/mitch/dev/karma-typescript/node_modules/log4js/lib/log4js.js' implicitly has an 'any' type.
  Try `npm install @types/log4js` if it exists or add a new declaration (.d.ts) file containing `declare module 'log4js';`
src/karma/framework.ts(2,24): error TS7016: Could not find a declaration file for module 'log4js'. '/home/mitch/dev/karma-typescript/node_modules/log4js/lib/log4js.js' implicitly has an 'any' type.
  Try `npm install @types/log4js` if it exists or add a new declaration (.d.ts) file containing `declare module 'log4js';`
src/karma/preprocessor.ts(3,24): error TS7016: Could not find a declaration file for module 'log4js'. '/home/mitch/dev/karma-typescript/node_modules/log4js/lib/log4js.js' implicitly has an 'any' type.
  Try `npm install @types/log4js` if it exists or add a new declaration (.d.ts) file containing `declare module 'log4js';`
src/karma/reporter.ts(6,24): error TS7016: Could not find a declaration file for module 'log4js'. '/home/mitch/dev/karma-typescript/node_modules/log4js/lib/log4js.js' implicitly has an 'any' type.
  Try `npm install @types/log4js` if it exists or add a new declaration (.d.ts) file containing `declare module 'log4js';`
src/shared/configuration.ts(2,25): error TS7016: Could not find a declaration file for module 'log4js'. '/home/mitch/dev/karma-typescript/node_modules/log4js/lib/log4js.js' implicitly has an 'any' type.
  Try `npm install @types/log4js` if it exists or add a new declaration (.d.ts) file containing `declare module 'log4js';`
src/shared/project.ts(7,24): error TS7016: Could not find a declaration file for module 'log4js'. '/home/mitch/dev/karma-typescript/node_modules/log4js/lib/log4js.js' implicitly has an 'any' type.
  Try `npm install @types/log4js` if it exists or add a new declaration (.d.ts) file containing `declare module 'log4js';`
```

I'm not sure how it passed for the most-recent commit of [monounity/karma-typescript](https://travis-ci.org/monounity/karma-typescript/builds/286838164?utm_source=github_status&utm_medium=notification). I can't just add `@types/log4js`, because it's not fully up-to-date

**Comment by mitchhentges** [comment]
@gempain are you seeing this issue locally on commit `https://github.com/monounity/karma-typescript/commit/dd891f988281501b06486fd8c725a256fd143fbe`?

**Comment by gempain** [comment]
@mitchhentges checking

**Comment by gempain** [comment]
Yup, I can confirm this locally. The CI build is broken as well. We need to install @types/log4js as explained in the error I think.

Sidenote: I use GitLab at work, it's much nicer to see which build failed and when. I miss the graph visualisation feature of Sourcetree on GitHub too :) It's a bit difficult to switch between Sourcetree, Travis, Veyor etc 😋

**Comment by erikbarke** [comment]
Log4js released v2.3.5 yesterday, the new version ships with its own typings and they seem kind of broken... I've tried locking to an earlier version which still uses the typings from @types/log4js, but I have to tread carefully here, the logging config object is shared between karma-typescript and the Transforms plugins, this will probably take some work 😕

**Comment by mitchhentges** [comment]
I can look into fixing this tonight (PDT) :)

**Comment by erikbarke** [comment]
The typings in log4js@2.3.5 seem broken beyond belief...

If you want to fix this @mitchhentges, you will also have to look at upgrading these plugins:
- [karma-typescript-angular2-transform](https://github.com/monounity/karma-typescript-angular2-transform)
- [karma-typescript-cssmodules-transform](https://github.com/monounity/karma-typescript-cssmodules-transform)
- [karma-typescript-es6-transform](https://github.com/monounity/karma-typescript-es6-transform)
- [karma-typescript-postcss-transform](https://github.com/monounity/karma-typescript-postcss-transform)

The plugins are fed the log level and the loggers config object from `karma` in [src/karma/framework.ts](https://github.com/monounity/karma-typescript/blob/master/src/karma/framework.ts#L28) and we could be looking at releasing all these plugins + karma-typescript as new major versions.

One quick fix would be locking to the last 1.x version, log4js@1.1.1.

**Comment by mitchhentges** [comment]
Locking to the last `1.x` version sounds like backing down from a challenge, let's update these puppies

**Comment by mitchhentges** [comment]
Update: [Karma still uses `log4js@^1.1.1`](https://github.com/karma-runner/karma/blob/master/package.json#L344). I think we should follow `karma` here, since the loggers are configured at the "karma level" of the config file, not within `karma-typescript`.

TL;DR `log4js@^1.1.1` it is

**Comment by erikbarke** [comment]
Awesome!

This means that whenever the Karma devs decide to upgrade to log4js 2.x, karma-typescript will have to support both the 1.x and the 2.x api to be able to run with newer and older versions of Karma. Hello, duck typing 😃

Closing since this was fixed in #197.

**Comment by mitchhentges** [comment]
When karma releases a new version, it might be worth releasing a new version of `karma-typescript` that only supports that new version, and have everything before it only support old `karma-typescript`, so there's not any overlap shenanigans.

Then:
* For new users, they'll install both new `karma` and new `karma-typescript`, which will work
* For existing users, when they update `karma`, they (should be) updating their plugins at the same time: I don't believe there's any guarantees that plugins will support varying major-release versions of typescript

**Comment by gempain** [comment]
Completely unrelated, but I have to say I enjoy reading both of you :smile: I'm learning quite a few new expressions and love the way you throw certain sentences :tada:
